arches/app/models/resource.py

```diff
@@ -307,24 +322,29 @@ def index(self, context=None):
                     context = {}
                 context["language"] = lang[0]
                 self.calculate_descriptors(context=context)
+
             datatype_factory = DataTypeFactory()
-            node_datatypes = {str(nodeid): datatype for nodeid, datatype in models.Node.objects.values_list("nodeid", "datatype")}
+
+            node_datatypes = {
+                str(nodeid): datatype for nodeid, datatype in ((k["nodeid"], k["datatype"]) for k in self.serialized_graph["nodes"])
+            }
             document, terms = self.get_documents_to_index(datatype_factory=datatype_factory, node_datatypes=node_datatypes, context=context)
             document["root_ontology_class"] = self.get_root_ontology()
             doc = JSONSerializer().serializeToPython(document)
             se.index_data(index=RESOURCES_INDEX, body=doc, id=self.pk)
             for term in terms:
                 se.index_data("terms", body=term["_source"], id=term["_id"])
 
-            celery_worker_running = task_management.check_if_celery_available()
+            if len(settings.ELASTICSEARCH_CUSTOM_INDEXES) > 0:
```

Comment:
Given that the performance cost is significant, maybe it makes sense to also check if any indexes have the "should_update_asynchronously" property?
```
indexes = settings.ELASTICSEARCH_CUSTOM_INDEXES
use_async = any([idx.get('should_update_asynchronously') for idx in indexes])
if indexes and use_async:
    celery_worker_running = task_management.check_if_celery_available()
```

Reply:
That would work as well, but honestly, I think we should be caching this value like we do in the DatatypeFactory because checking this every time a resource uses a custom index is still overkill especially when considering bulk upload/indexing. 
That would probably be better to handle in another ticket IMO.